funnel/models/event.py

```diff
@@ -143,5 +202,32 @@ class SyncTicket(BaseMixin, db.Model):
     ticket_client_id = db.Column(db.Integer, db.ForeignKey('ticket_client.id'), nullable=False)
     ticket_client = db.relationship(TicketClient,
         backref=db.backref('sync_tickets', cascade='all, delete-orphan'))
-
     __table_args__ = (db.UniqueConstraint('proposal_space_id', 'order_no', 'ticket_no'),)
+
+    @classmethod
+    def get(cls, space, order_no, ticket_no):
+        return cls.query.filter_by(proposal_space=space, order_no=order_no, ticket_no=ticket_no).one_or_none()
+
+    @classmethod
+    def upsert(cls, space, order_no, ticket_no, participant, ticket_client, ticket_type):
+        ticket = cls.get(space, order_no, ticket_no)
+
+        if not ticket:
+            ticket = SyncTicket(proposal_space=space, order_no=order_no,
+                             ticket_no=ticket_no, ticket_type=ticket_type,
+                             participant=participant, ticket_client=ticket_client)
+            db.session.add(ticket)
+        else:
+            # Transfer ticket
+            if ticket.participant is not participant:
+                ticket.participant.update_events(ticket.ticket_type.events, cancel=True)
+                ticket.participant = participant
+
+        ticket.participant.update_events(ticket.ticket_type.events)
```

Comment:
Upsert on a SyncTicket instance modifies a Participant instance? That's not good. This code doesn't belong here.